File: FarmaciasDeGuardia/Services/CoordinateCache.swift
import Foundation
import CoreLocation

/// Manages persistent coordinate caching for pharmacy addresses
class CoordinateCache {
    private static let userDefaults = UserDefaults.standard
    private static let cacheKey = "pharmacy_coordinates_cache"
    private static let cacheVersionKey = "coordinate_cache_version"
    private static let currentCacheVersion = 1
    
    /// Get cached coordinates for an address
    static func getCoordinates(for address: String) -> CLLocation? {
        guard let cacheData = userDefaults.data(forKey: cacheKey),
              let cache = try? JSONDecoder().decode([String: CachedCoordinate].self, from: cacheData),
              let cached = cache[address],
              !cached.isExpired else {
            return nil
        }
        
        return CLLocation(latitude: cached.latitude, longitude: cached.longitude)
    }
    
    /// Cache coordinates for an address
    static func setCoordinates(_ location: CLLocation, for address: String) {
        var cache = loadCache()
        cache[address] = CachedCoordinate(
            latitude: location.coordinate.latitude,
            longitude: location.coordinate.longitude,
            timestamp: Date()
        )
        saveCache(cache)
    }
    
    /// Clear expired cache entries
    static func cleanupExpiredEntries() {
        var cache = loadCache()
        let originalCount = cache.count
        
        cache = cache.filter { !$0.value.isExpired }
        
        if cache.count != originalCount {
            saveCache(cache)
            DebugConfig.debugPrint("🧹 Cleaned up \(originalCount - cache.count) expired coordinate cache entries")
        }
    }
    
    /// Clear all cached coordinates
    static func clearAll() {
        userDefaults.removeObject(forKey: cacheKey)
        DebugConfig.debugPrint("🗑️ Cleared all coordinate cache")
    }
    
    /// Get cache statistics
    static func getCacheStats() -> (count: Int, size: String) {
        let cache = loadCache()
        let data = userDefaults.data(forKey: cacheKey) ?? Data()
        let sizeKB = Double(data.count) / 1024.0
        return (count: cache.count, size: String(format: "%.1f KB", sizeKB))
    }
    
    private static func loadCache() -> [String: CachedCoordinate] {
        // Check cache version
        let savedVersion = userDefaults.integer(forKey: cacheVersionKey)
        if savedVersion != currentCacheVersion {
            // Clear cache if version mismatch
            userDefaults.removeObject(forKey: cacheKey)
            userDefaults.set(currentCacheVersion, forKey: cacheVersionKey)
            return [:]
        }
        
        guard let data = userDefaults.data(forKey: cacheKey),
              let cache = try? JSONDecoder().decode([String: CachedCoordinate].self, from: data) else {
            return [:]
        }
        
        return cache
    }
    
    private static func saveCache(_ cache: [String: CachedCoordinate]) {
        if let data = try? JSONEncoder().encode(cache) {
            userDefaults.set(data, forKey: cacheKey)
        }
    }
}

private struct CachedCoordinate: Codable {
    let latitude: Double
    let longitude: Double
    let timestamp: Date
    
    // Cache expires after 30 days
    var isExpired: Bool {
        Date().timeIntervalSince(timestamp) > 30 * 24 * 60 * 60
    }
}

File: android/app/src/main/java/com/github/bfollon/farmaciasdeguardiaensegovia/services/CoordinateCache.kt
/*
 * Copyright (C) 2025  Bruno Follon (@bFollon)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.github.bfollon.farmaciasdeguardiaensegovia.services

import android.content.Context
import android.content.SharedPreferences
import android.location.Location
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Manages persistent coordinate caching for pharmacy addresses
 * Equivalent to iOS CoordinateCache using UserDefaults
 */
object CoordinateCache {
    
    private const val CACHE_KEY = "pharmacy_coordinates_cache"
    private const val CACHE_VERSION_KEY = "coordinate_cache_version"
    private const val CURRENT_CACHE_VERSION = 1
    private const val CACHE_EXPIRY_DAYS = 30L // 30 days expiry (same as iOS)
    
    private lateinit var sharedPreferences: SharedPreferences
    private val json = Json { ignoreUnknownKeys = true }
    
    /**
     * Initialize the cache with application context
     */
    fun initialize(context: Context) {
        sharedPreferences = context.getSharedPreferences("coordinate_cache", Context.MODE_PRIVATE)
        
        // Check cache version and clear if outdated
        val currentVersion = sharedPreferences.getInt(CACHE_VERSION_KEY, 0)
        if (currentVersion != CURRENT_CACHE_VERSION) {
            DebugConfig.debugPrint("🔄 Cache version mismatch, clearing coordinate cache")
            clearAll()
            sharedPreferences.edit()
                .putInt(CACHE_VERSION_KEY, CURRENT_CACHE_VERSION)
                .apply()
        }
    }
    
    /**
     * Get cached coordinates for an address
     */
    fun getCoordinates(address: String): Location? {
        if (!::sharedPreferences.isInitialized) {
            DebugConfig.debugPrint("❌ CoordinateCache not initialized")
            return null
        }
        
        val cacheJson = sharedPreferences.getString(CACHE_KEY, null) ?: return null
        
        return try {
            val cache = json.decodeFromString<Map<String, CachedCoordinate>>(cacheJson)
            val cached = cache[address]
            
            if (cached != null && !cached.isExpired) {
                Location("cache").apply {
                    latitude = cached.latitude
                    longitude = cached.longitude
                }
            } else {
                null
            }
        } catch (exception: Exception) {
            DebugConfig.debugPrint("❌ Failed to decode coordinate cache: ${exception.message}")
            null
        }
    }
    
    /**
     * Cache coordinates for an address
     */
    fun setCoordinates(location: Location, address: String) {
        if (!::sharedPreferences.isInitialized) {
            DebugConfig.debugPrint("❌ CoordinateCache not initialized")
            return
        }
        
        val cache = loadCache().toMutableMap()
        cache[address] = CachedCoordinate(
            latitude = location.latitude,
            longitude = location.longitude,
            timestamp = System.currentTimeMillis()
        )
        saveCache(cache)
    }
    
    /**
     * Clear expired cache entries
     */
    fun cleanupExpiredEntries() {
        if (!::sharedPreferences.isInitialized) {
            DebugConfig.debugPrint("❌ CoordinateCache not initialized")
            return
        }
        
        val cache = loadCache().toMutableMap()
        val originalCount = cache.size
        
        cache.entries.removeAll { it.value.isExpired }
        
        if (cache.size != originalCount) {
            saveCache(cache)
            DebugConfig.debugPrint("🧹 Cleaned up ${originalCount - cache.size} expired coordinate cache entries")
        }
    }
    
    /**
     * Clear all cached coordinates
     */
    fun clearAll() {
        if (!::sharedPreferences.isInitialized) {
            DebugConfig.debugPrint("❌ CoordinateCache not initialized")
            return
        }
        
        sharedPreferences.edit()
            .remove(CACHE_KEY)
            .apply()
        DebugConfig.debugPrint("🗑️ Cleared all coordinate cache")
    }
    
    /**
     * Get cache statistics
     */
    fun getCacheStats(): Pair<Int, String> {
        if (!::sharedPreferences.isInitialized) {
            return 0 to "0B"
        }
        
        val cache = loadCache()
        val count = cache.size
        val sizeEstimate = count * 150 // Rough estimate: 150 bytes per entry
        val sizeString = when {
            sizeEstimate < 1024 -> "${sizeEstimate}B"
            sizeEstimate < 1024 * 1024 -> "${sizeEstimate / 1024}KB"
            else -> "${sizeEstimate / (1024 * 1024)}MB"
        }
        return count to sizeString
    }
    
    /**
     * Load cache from SharedPreferences
     */
    private fun loadCache(): Map<String, CachedCoordinate> {
        val cacheJson = sharedPreferences.getString(CACHE_KEY, null) ?: return emptyMap()
        
        return try {
            json.decodeFromString<Map<String, CachedCoordinate>>(cacheJson)
        } catch (exception: Exception) {
            DebugConfig.debugPrint("❌ Failed to load coordinate cache: ${exception.message}")
            emptyMap()
        }
    }
    
    /**
     * Save cache to SharedPreferences
     */
    private fun saveCache(cache: Map<String, CachedCoordinate>) {
        try {
            val cacheJson = json.encodeToString(cache)
            sharedPreferences.edit()
                .putString(CACHE_KEY, cacheJson)
                .apply()
        } catch (exception: Exception) {
            DebugConfig.debugPrint("❌ Failed to save coordinate cache: ${exception.message}")
        }
    }
}

/**
 * Represents a cached coordinate with expiration
 */
@Serializable
data class CachedCoordinate(
    val latitude: Double,
    val longitude: Double,
    val timestamp: Long
) {
    /**
     * Check if this cached coordinate has expired
     */
    val isExpired: Boolean
        get() {
            val expiryTime = timestamp + (CACHE_EXPIRY_DAYS * 24 * 60 * 60 * 1000)
            return System.currentTimeMillis() > expiryTime
        }
    
    companion object {
        private const val CACHE_EXPIRY_DAYS = 30L // 30 days expiry
    }
}
